File: src/api/memories.py
import uuid
from typing import Literal

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession
from src.api.documents import get_db_session
from src.dependencies import get_memory_service, get_profile_service_instance

router = APIRouter(prefix="/api/memory", tags=["memory"])


class RememberRequest(BaseModel):
    content: str
    source: str = "explicit"


class ObserveRequest(BaseModel):
    conversation: str


@router.post("/remember")
async def remember(
    body: RememberRequest,
    session: AsyncSession = Depends(get_db_session),
):
    service = get_memory_service(session)
    result = await service.remember(session, body.content, source=body.source)
    await session.commit()
    return {"facts_extracted": result.facts_extracted, "memories_updated": result.memories_updated}


@router.post("/observe")
async def observe(
    body: ObserveRequest,
    session: AsyncSession = Depends(get_db_session),
):
    service = get_memory_service(session)
    result = await service.observe(session, body.conversation)
    await session.commit()
    return {"facts_extracted": result.facts_extracted, "memories_updated": result.memories_updated}


@router.get("/list")
async def list_memories(
    session: AsyncSession = Depends(get_db_session),
    category: Literal["research", "reminder", "thought", "decision", "preference"] | None = Query(default=None),
):
    service = get_memory_service(session)
    memories = await service.list_active(session, category=category)
    return [
        {
            "id": str(m.id),
            "content": m.content,
            "source": m.source,
            "category": m.category,
            "project": m.project,
            "relation": m.relation,
            "created_at": m.created_at.isoformat() if m.created_at else None,
        }
        for m in memories
    ]


@router.get("/context")
async def context(
    session: AsyncSession = Depends(get_db_session),
):
    service = get_memory_service(session)
    profile_service = get_profile_service_instance()
    memories = await service.list_active(session)
    profile = await profile_service.build_profile(memories)
    return {"static": profile.static, "dynamic": profile.dynamic, "raw_count": profile.raw_count}


@router.delete("/{memory_id}")
async def forget_memory(
    memory_id: uuid.UUID,
    session: AsyncSession = Depends(get_db_session),
):
    service = get_memory_service(session)
    ok = await service.forget_memory(session, memory_id)
    await session.commit()
    if not ok:
        raise HTTPException(status_code=404, detail="Memory not found")
    return {"status": "deleted"}

File: rust/src/api/memories.rs
use axum::extract::{Path, Query, State};
use axum::http::StatusCode;
use axum::routing::{delete, get, post};
use axum::{Json, Router};
use serde::{Deserialize, Serialize};

use crate::db::repositories::memories::{Memory, MemoryRepository};
use crate::error::AppError;

use super::state::AppState;

#[derive(Deserialize)]
pub struct RememberRequest {
    pub content: String,
    #[serde(default = "default_source")]
    pub source: String,
}

fn default_source() -> String {
    "explicit".to_string()
}

#[derive(Deserialize)]
pub struct ObserveRequest {
    pub conversation: String,
}

#[derive(Deserialize)]
pub struct ListQuery {
    pub category: Option<String>,
}

#[derive(Serialize)]
pub struct RememberResponse {
    pub facts_extracted: usize,
    pub memories_updated: usize,
}

#[derive(Serialize)]
pub struct MemoryItem {
    pub id: String,
    pub content: String,
    pub source: String,
    pub category: Option<String>,
    pub project: Option<String>,
    pub relation: Option<String>,
    pub created_at: String,
}

#[derive(Serialize)]
pub struct ProfileResponse {
    #[serde(rename = "static")]
    pub static_summary: String,
    #[serde(rename = "dynamic")]
    pub dynamic_summary: String,
    pub raw_count: usize,
}

pub fn router() -> Router<AppState> {
    Router::new()
        .route("/api/memory/remember", post(remember_handler))
        .route("/api/memory/observe", post(observe_handler))
        .route("/api/memory/list", get(list_memories_handler))
        .route("/api/memory/context", get(context_handler))
        .route("/api/memory/:id", delete(forget_handler))
}

async fn remember_handler(
    State(state): State<AppState>,
    Json(body): Json<RememberRequest>,
) -> Result<Json<RememberResponse>, AppError> {
    let pool = state.pool.clone();
    let memory_svc = state.memory_svc.clone();
    let content = body.content.clone();
    let source = body.source.clone();

    let result = tokio::task::spawn_blocking(move || {
        let conn = pool.get().map_err(AppError::Pool)?;
        tokio::runtime::Handle::current()
            .block_on(async move { memory_svc.remember(&conn, &content, &source).await })
            .map_err(|e| AppError::Llm(e.to_string()))
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    Ok(Json(RememberResponse {
        facts_extracted: result.facts_extracted,
        memories_updated: result.memories_updated,
    }))
}

async fn observe_handler(
    State(state): State<AppState>,
    Json(body): Json<ObserveRequest>,
) -> Result<Json<RememberResponse>, AppError> {
    let pool = state.pool.clone();
    let memory_svc = state.memory_svc.clone();
    let conversation = body.conversation.clone();

    let result = tokio::task::spawn_blocking(move || {
        let conn = pool.get().map_err(AppError::Pool)?;
        tokio::runtime::Handle::current()
            .block_on(async move { memory_svc.observe(&conn, &conversation).await })
            .map_err(|e| AppError::Llm(e.to_string()))
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    Ok(Json(RememberResponse {
        facts_extracted: result.facts_extracted,
        memories_updated: result.memories_updated,
    }))
}

async fn list_memories_handler(
    State(state): State<AppState>,
    Query(params): Query<ListQuery>,
) -> Result<Json<Vec<MemoryItem>>, AppError> {
    let pool = state.pool.clone();
    let category = params.category.clone();

    let memories = tokio::task::spawn_blocking(move || -> Result<Vec<Memory>, AppError> {
        let conn = pool.get().map_err(AppError::Pool)?;
        let repo = MemoryRepository::new(&conn);
        let all = repo.get_all_active().map_err(AppError::Db)?;

        // Filter by category if provided
        if let Some(cat) = &category {
            Ok(all
                .into_iter()
                .filter(|m| m.category.as_deref() == Some(cat.as_str()))
                .collect())
        } else {
            Ok(all)
        }
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    let items = memories
        .into_iter()
        .map(|m| MemoryItem {
            id: m.id,
            content: m.content,
            source: m.source,
            category: m.category,
            project: m.project,
            relation: m.relation,
            created_at: m.created_at,
        })
        .collect();

    Ok(Json(items))
}

async fn context_handler(
    State(state): State<AppState>,
) -> Result<Json<ProfileResponse>, AppError> {
    let pool = state.pool.clone();

    // Get all active memories (blocking)
    let memories = tokio::task::spawn_blocking(move || -> Result<Vec<Memory>, AppError> {
        let conn = pool.get().map_err(AppError::Pool)?;
        let repo = MemoryRepository::new(&conn);
        repo.get_all_active().map_err(AppError::Db)
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    // Build profile (async, calls LLM) — profile_svc is Arc<ProfileService> which is Send+Sync
    let profile_svc = state.profile_svc.clone();
    let profile = profile_svc
        .build_profile(&memories)
        .await
        .map_err(|e| AppError::Llm(e.to_string()))?;

    Ok(Json(ProfileResponse {
        static_summary: profile.static_summary,
        dynamic_summary: profile.dynamic_summary,
        raw_count: profile.raw_count,
    }))
}

async fn forget_handler(
    State(state): State<AppState>,
    Path(memory_id): Path<String>,
) -> Result<Json<serde_json::Value>, AppError> {
    let pool = state.pool.clone();
    let memory_svc = state.memory_svc.clone();
    let memory_id_clone = memory_id.clone();

    let found = tokio::task::spawn_blocking(move || {
        let conn = pool.get().map_err(AppError::Pool)?;
        tokio::runtime::Handle::current()
            .block_on(async move { memory_svc.forget(&conn, &memory_id_clone).await })
            .map_err(|e| AppError::Llm(e.to_string()))
    })
    .await
    .map_err(|e| AppError::Llm(format!("task join error: {e}")))??;

    if found {
        Ok(Json(serde_json::json!({"status": "deleted"})))
    } else {
        Err(AppError::NotFound(format!("memory {memory_id}")))
    }
}
